service/worker/scheduler/workflow.go

```diff
@@ -423,8 +428,20 @@ func (s *scheduler) sleep(nextWakeup time.Time) {
 		if !s.tweakables.AllowZeroSleep && sleepTime <= 0 {
 			sleepTime = time.Second
 		}
-		tmr := workflow.NewTimer(s.ctx, sleepTime)
-		sel.AddFuture(tmr, func(_ workflow.Future) {})
+		// A previous version of this workflow always created a new timer here, which is wasteful.
+		// We can reuse a previous timer if we have the same deadline and it didn't fire yet.
+		if s.tweakables.ReuseTimer {
+			if s.currentTimer == nil || !s.currentTimerDeadline.Equal(nextWakeup) {
+				s.currentTimer = workflow.NewTimer(s.ctx, sleepTime)
```

Comment:
This doesn't attempt to cancel a previous timer if the deadline changed (this would only happen on update schedule). If we think that's worth it I can add that too